include/Zydis/Encoder.h

```diff
@@ -0,0 +1,422 @@
+        ZYAN_BITS_TO_REPRESENT(ZYDIS_ENCODABLE_ENCODING_MAX_VALUE)
+} ZydisEncodableEncoding;
+
+/**
+ * Defines the `ZydisEncodablePrefix` enum.
+ */
+typedef enum ZydisEncodablePrefix_
+{
+    ZYDIS_ENCODABLE_PREFIX_NONE                     = 0x00000000,
+    ZYDIS_ENCODABLE_PREFIX_LOCK                     = 0x00000001,
+    ZYDIS_ENCODABLE_PREFIX_REP                      = 0x00000002,
+    ZYDIS_ENCODABLE_PREFIX_REPE                     = 0x00000004,
+    ZYDIS_ENCODABLE_PREFIX_REPZ                     = ZYDIS_ENCODABLE_PREFIX_REPE,
+    ZYDIS_ENCODABLE_PREFIX_REPNE                    = 0x00000008,
+    ZYDIS_ENCODABLE_PREFIX_REPNZ                    = ZYDIS_ENCODABLE_PREFIX_REPNE,
+    ZYDIS_ENCODABLE_PREFIX_BND                      = 0x00000010,
+    ZYDIS_ENCODABLE_PREFIX_XACQUIRE                 = 0x00000020,
+    ZYDIS_ENCODABLE_PREFIX_XRELEASE                 = 0x00000040,
+    ZYDIS_ENCODABLE_PREFIX_BRANCH_NOT_TAKEN         = 0x00000080,
+    ZYDIS_ENCODABLE_PREFIX_BRANCH_TAKEN             = 0x00000100,
+    ZYDIS_ENCODABLE_PREFIX_NOTRACK                  = 0x00000200,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_CS               = 0x00000400,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_SS               = 0x00000800,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_DS               = 0x00001000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_ES               = 0x00002000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_FS               = 0x00004000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS               = 0x00008000,
+
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_MASK             = (ZYDIS_ENCODABLE_PREFIX_SEGMENT_CS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_SS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_DS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_ES |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_FS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS),
+
+    /**
+     * Maximum value of this enum.
+     */
+    ZYDIS_ENCODABLE_PREFIX_MAX_VALUE                = (ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS |
+                                                       (ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS - 1)),
+    /**
+     * The minimum number of bits required to represent all values of this enum.
+     */
+    ZYDIS_ENCODABLE_PREFIX_REQUIRED_BITS            =
+        ZYAN_BITS_TO_REPRESENT(ZYDIS_ENCODABLE_PREFIX_MAX_VALUE)
+} ZydisEncodablePrefix;
+
+/**
+ * Defines the `ZydisEncodableBranchType` enum.
+ */
+typedef enum ZydisEncodableBranchType_
+{
+    ZYDIS_ENCODABLE_BRANCH_TYPE_NONE,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_SHORT,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_NEAR16,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_NEAR32,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_NEAR64,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_FAR16,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_FAR32,
+    ZYDIS_ENCODABLE_BRANCH_TYPE_FAR64,
```

Comment:
From encoder's perspective that's a desirable change. I needed a type that essentially encodes a tuple (branch_type, branch_immediate_value_size) hence `ZydisEncodableBranchType` was born. Merging these types would make translation between them unnecessary and that's a welcome change.

Now let's look at decoder impact. In decoder `branch_type` provides branch type ONLY and if anyone needs information about its size it must be extracted separately. Only downside is that if you are interested in just branch type e.g. `NEAR` you would need to test for `NEAR16 | NEAR32 | NEAR64`. Is that a big deal? Probably not.

Only piece of code that uses `branch_type` in decoder is this one:
```
                instruction->meta.branch_type = definition->branch_type;
                ZYAN_ASSERT((instruction->meta.branch_type == ZYDIS_BRANCH_TYPE_NONE) ||
                        ((instruction->meta.category == ZYDIS_CATEGORY_CALL) ||
                         (instruction->meta.category == ZYDIS_CATEGORY_COND_BR) ||
                         (instruction->meta.category == ZYDIS_CATEGORY_UNCOND_BR) ||
                         (instruction->meta.category == ZYDIS_CATEGORY_RET)));
```
So no major problems here as well. I'd say I'm in favor of such change for the sake of simplicity. However I will need to count on you guys because this requires generator change (`definition->branch_type` as seen above).

Reply:
I would be fine with converting the decoder flag enum to flags adding a combined mask for backwards compatibility.

We just have to check if the decoder always easily knows the size (e.g. even in minimal mode). The tables only save the NEAR/FAR/SHORT because some of the instructions are affected by the 66h prefix.

Reply:
Note that this would still be a breaking change even if defining it as flags, because for the mask to work, you have to do `& BRANCH_TYPE` whereas existing code will most likely currently use `== BRANCH_TYPE`. 

Unrelated to the comment above, if we do flags, instead of having a mask, we could also represent it like this:

```c
enum ZydisBranchFlags {
  ZYDIS_BRANCH_FLAGS_16 = 1 << 0,
  ZYDIS_BRANCH_FLAGS_32 = 1 << 1,
  ZYDIS_BRANCH_FLAGS_64 = 1 << 2,

  ZYDIS_BRANCH_FLAGS_SHORT = 1 << 3,
  ZYDIS_BRANCH_FLAGS_NEAR = 1 << 4,
  ZYDIS_BRANCH_FLAGS_FAR = 1 << 5,
};
```

The drawback with this approach is that it would allow for nonsensical combinations of flags like short+64.